### PR TITLE
Anchor v1

### PR DESCRIPTION
**Pull request by nourharidy**
**Architecture overview for reviewers:**

Anchor is a fork of Compound. The main compound contracts (CToken, Comptroller, etc are not changed). The only change to the Compound protocol is the PriceOracle implementation which plugs into Chainlink feeds.

The main goal of the Anchor project is to re-use Compound's loan and debt management plumbing for the purpose of synthetic asset creation (Turning Compound into MakerDAO/Synthetix). This is achieved by creating an ERC20 token for each synth, minting it and supplying it exclusively to the protocol which will lend it to others in exchange for collateral.

**This achieves essentially the same architecture as MakerDAO but with few important differences:**
- The max supply of an Anchor-listed synth at any given point in time is defined by the amount minted and supplied to the protocol, as opposed to being minted on the fly in the case of Maker or Synthetix
- The borrowing interest rate (aka Stability Fee in the case of Maker) is dynamic based on the current interest rate model (as opposed to a preset rate defined by gov in the case of Maker). This also means that interest rate is a function of minted synth supply supplied to Anchor by gov and the borrowed supply, according to Compound's current interest rate model (which is subject to future experimentation). Governance can raise borrowing rates indirectly by monetary `contraction` by redeeming and burning some of the synth on Anchor, or lower rates by monetary `expansion` by minting and supplying more tokens to the protocol. This functionality is controlled by the `Fed` contract whose main mission is to use these powers to stabilize the peg.
- Interest generated from borrowing is shared between the governance (via the `Fed`) as the initial supplier and other synth holders who choose to re-deposit/supply their synths to the protocol. This means that all synths are natively yield-bearing. Governance can choose to lower the share of other suppliers interest by increasing Compound's `reserveFactor` variable.
- In the future, borrowing ETH using Dola as collateral will be allowed, meaning that those who supply ETH to use as collateral will also earn interest on their collateral (unlike Maker).
- In the future, by current design, different synths can be used to collateralize other synths (unlike Synthetix), which allows users to enter some unique leveraged positions.

**The main new contracts introduced are the following:**
- `ERC20`: Represents a synth token instance (e.g. Dola). Supports multiple minters. First minter will be the Fed. In the future, other minters can be added to allow for other functionalities such as 0-collateral lending.
- `Fed`: Each synth has its own Fed instance. The Fed is the middleware between the synth contract and Anchor and controls the monetary `expansion` and `contraction` of the synth on Anchor. Each Fed has a `chair` (owner), who is able to directly control monetary policy without having direct access to funds in transit.
- `Oracle`: A global oracle contract compatible with Compound's oracle interface. It is the middleware that transforms Chainlink price feed oracle format to Compound's. It also support setting fixed prices for certain assets (e.g. Dola which is set to $1).

**Initial launch phase:**
The main goal during the launch is to be as lean as possible. Anchor is very ambitious project but it must first be proven in its most simple form:
- In this phase, only one synth (Dola, pegged to $1) and one collateral (ETH) are supported.
- Dola can only be borrowed using ETH as collateral (and not the other way around, ETH borrowing is paused).
- Borrowing limit will be very uncompetitive at 50% limit (200% collateralization/liquidation threshold) to lower initial risk of insolvency.

As a reviewer, you should be verifying that the architecture and all features above are represented correctly in the code base. This not only includes the contracts source code, but also the deployment scripts located at `deploy/` directory. There is also a single unit test for the `Oracle` contract in the `test/` directory given the math involved unlike other contracts. There's no need to fully review the entirety of the Compound code base. However, it is important to ensure that these components will function as intended combined with our code. Only `Comptroller.sol`, `Unitroller.sol`, `CErc20Immutable.sol`, `CEther.sol`, `JumpRateModelV2.sol` and `WhitePaperInterestRateModel.sol` from Compound's code are used in our protocol in addition to our own contracts.

Let me know if there's anything I can help with.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,275 +1,18 @@
-[![CircleCI](https://circleci.com/gh/compound-finance/compound-protocol.svg?style=svg&circle-token=5ed19932325c559a06f71f87d69012aedd2cf3fb)](https://circleci.com/gh/compound-finance/compound-protocol) [![codecov](https://codecov.io/gh/compound-finance/compound-protocol/branch/master/graph/badge.svg?token=q4UvsvVzOX)](https://codecov.io/gh/compound-finance/compound-protocol)
-
-Compound Protocol
+Anchor Protocol
```
Comment (0xKiwi) [inline README.md:1]: Would be nice to add something like the PR description here.

File: contracts/ComptrollerG6.sol [hunk 2]
```diff
@@ -1280,7 +1280,7 @@ contract ComptrollerG6 is ComptrollerV5Storage, ComptrollerInterface, Comptrolle
      * @return The amount of COMP which was NOT transferred to the user
      */
     function grantCompInternal(address user, uint amount) internal returns (uint) {
-        Comp comp = Comp(getCompAddress());
+        INV comp = INV(getCompAddress());
```
Comment (0xKiwi) [inline contracts/ComptrollerG6.sol:1283]: Is this functionality needed? Isn't this for COMP rewards? Are we mirroring that with INV?

File: contracts/ERC20.sol [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)'),
+                keccak256(bytes(name)),
+                keccak256(bytes('1')),
+                chainId,
+                address(this)
+            )
+        );
+    }
+
+    function changeOperator(address operator_) public onlyOperator {
```
Comment (0xKiwi) [inline contracts/ERC20.sol:53]: No safer function for transfering here? Like `setPendingOperator` and `claimOperator`?

File: contracts/Fed.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    function resign() public {
+        require(msg.sender == chair, "ONLY CHAIR");
+        chair = address(0);
+    }
+
+    function expansion(uint amount) public {
+        require(msg.sender == chair, "ONLY CHAIR");
+        underlying.mint(address(this), amount);
```
Comment (0xKiwi) [inline contracts/Fed.sol:45]: Underlying here would be DOLA?

File: contracts/Oracle.sol [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    constructor() public {
+        owner = msg.sender;
+    }
+
+    function changeOwner(address owner_) public onlyOwner {
+        owner = owner_;
+    }
+
+    function setFeed(CToken cToken_, address feed_, uint8 tokenDecimals_) public onlyOwner {
+        feeds[address(cToken_)] = FeedData(feed_, tokenDecimals_);
```
Comment (0xKiwi) [inline contracts/Oracle.sol:36]: The feed is on the cToken? Not the underlying?

File: contracts/ComptrollerG3.sol [hunk 6]
```diff
@@ -1246,7 +1246,7 @@ contract ComptrollerG3 is ComptrollerV3Storage, ComptrollerInterface, Comptrolle
      */
     function transferComp(address user, uint userAccrued, uint threshold) internal returns (uint) {
         if (userAccrued >= threshold && userAccrued > 0) {
-            Comp comp = Comp(getCompAddress());
```
Comment (0xKiwi) [inline contracts/ComptrollerG3.sol:1249]: Can all these old Comptroller contracts be discarded or no?

File: contracts/Fed.sol (hunk 4 above)
Comment (niktrix) [reply contracts/Fed.sol:45]: underlying is dola here,         underlying = ERC20(ctoken_.underlying());

File: contracts/ComptrollerG6.sol (hunk 2 above)
Comment (nourharidy) [reply contracts/ComptrollerG6.sol:1283]: yes we are

File: contracts/ERC20.sol (hunk 3 above)
Comment (nourharidy) [reply contracts/ERC20.sol:53]: will add

File: contracts/Fed.sol (hunk 4 above)
Comment (nourharidy) [reply contracts/Fed.sol:45]: Each synth has a Fed. Dola is only the first synth. We get the synth address from its cToken

File: contracts/Oracle.sol (hunk 5 above)
Comment (nourharidy) [reply contracts/Oracle.sol:36]: it's the same. Just makes it easier to add ETH which is not a token

File: contracts/ComptrollerG3.sol (hunk 6 above)
Comment (nourharidy) [reply contracts/ComptrollerG3.sol:1249]: We can do some refactoring after launch. Let's do as few changes as possible for now

File: README.md (hunk 1 above)
Comment (nourharidy) [reply README.md:1]: done. Please resolve if ok

File: contracts/ERC20.sol (hunk 3 above)
Comment (nourharidy) [reply contracts/ERC20.sol]: done. Please resolve

File: contracts/Fed.sol (hunk 4 above)
Comment (0xKiwi) [reply contracts/Fed.sol:45]: Are we gonna keep the name `cToken` for this?

Comment (nourharidy) [reply contracts/Fed.sol:45]: idk tbh. We should come up with another prefix

**Pull request merged**